vim: Fix bug where `d . .` freezes the editor

AidanV opened the pull request:
This bug seems to be caused by pushing an operator (i.e. `d`) followed by a repeat (i.e. `.`)  so the recording includes the push operator and the repeat. When this is repeated (i.e. `.`) it causes an infinite loop. 

This change fixes this bug by pushing a ClearOperator action if there is an ongoing recording when repeat is called.

Release Notes:

- Fixed bug where pressing `d . .` in Vim mode would freeze the editor.

AidanV committed:
Fix bug for repeat in recording

dinocosta committed:
test(vim): add test for changes done to repeat

dinocosta commented:
Tested this a little bit and it actually already seems to be doing somewhat the same as NeoVim when it comes to resetting the operator. 

However, it seems that this breaks the recorded action too. For example, trying these actions leads to different outcomes in NeoVim and Zed:

1. Delete a word with `dw`
2. Confirm that you're able to use `2.` to delete 2 words
3. Confirm that using `2d.` does nothing
4. Attempt to use the repeat operation (`.`)
    - In Zed nothing will happen, I suspect we're deleting the recorded actions?
    - In NeoVim, the recorded action continues working as expected
    
Let me know if you want to try a look @AidanV . Can also try checking this next week, otherwise will merge these changes Wednesday morning, to ensure these land in the next Preview 👍

AidanV committed:
test(vim): add neovim test for `d . .`

AidanV committed:
fix(vim): `d d d . .` deletes two lines

AidanV committed:
test(vim): Add `2 d .` test

AidanV committed:
fix(vim): Fix `2 d .` bug

AidanV committed:
test(vim): Add `2 d d 5 d . .` test

AidanV committed:
fix(vim): Separate recording and recorded counts

AidanV commented:
I added a test and a fix for the bug you pointed out. 

I also added a test and a fix to catch a bug I found where if we have `2 d d` followed by a different count that is dropped (such as `5 d .`) then the next `.` would only delete one line instead of the expected two lines.

In order to fix this I added a `recording_count` field to the Vim globals. This new variable mirrors the relationship that `recording_actions` and `recorded_actions` have with `recording_count` and `recorded_count`.

I can't find any more bugs, but I am sure they exist :).

Thank you for you help on this @dinocosta!

dinocosta commented:
Thank you for taking the time to update this @AidanV ! I've tested and confirm that it seems everything is working as expected 🎉

I'm holding off on merging this until after we have a new Preview release out, so we can have this in Nightly for a week and have more people experiencing it. I'll also see if I can take some time to revisit the approach, as I'm suspicious that we should be able to get away with just the `recorded_count` field, ensuring that we don't reset it when we start recording, but only set it when we stop recording 🤔

AidanV commented:
@dinocosta gentle reminder to merge this :)
